workflow-runtime/src/commonTest/kotlin/com/squareup/workflow1/RenderWorkflowInTest.kt

```diff
@@ -377,85 +341,168 @@ class RenderWorkflowInTest {
-          ) {
-            action("") {
-              state = it
-              setOutput(it)
+      runTest(UnconfinedTestDispatcher()) {
+        val trigger = Channel<String>()
+        val workflow = Workflow.stateful<String, String, String>(
+          initialState = "initial",
+          render = { renderState ->
+            runningWorker(
+              trigger.receiveAsFlow()
+                .asWorker()
+            ) {
+              action("") {
+                state = it
+                setOutput(it)
+              }
             }
+            renderState
           }
-          renderState
+        )
+
+        val receivedOutputs = mutableListOf<String>()
+        lateinit var renderings: StateFlow<RenderingAndSnapshot<String>>
+        renderings = renderWorkflowIn(
+          workflow = workflow,
+          scope = backgroundScope,
+          props = MutableStateFlow(Unit),
+          runtimeConfig = runtimeConfig,
+          workflowTracer = workflowTracer,
+        ) { it: String ->
+          receivedOutputs += it
+          // The value of the updated rendering has already been set by the time onOutput is
+          // called
+          assertEquals(it, renderings.value.rendering)
         }
-      )
-
-      val emittedRenderings = mutableListOf<String>()
-      val receivedOutputs = mutableListOf<String>()
-      val renderings = renderWorkflowIn(
-        workflow = workflow,
-        scope = testScope,
-        props = MutableStateFlow(Unit),
-        runtimeConfig = runtimeConfig,
-        workflowTracer = workflowTracer,
-      ) { it: String ->
-        receivedOutputs += it
-        assertTrue(emittedRenderings.contains(it))
+        assertTrue(receivedOutputs.isEmpty())
+
+        assertTrue(trigger.trySend("foo").isSuccess)
+        assertEquals(listOf("foo"), receivedOutputs)
       }
-      assertTrue(receivedOutputs.isEmpty())
+    }
+  }
 
-      val scope = CoroutineScope(Unconfined)
-      scope.launch {
-        renderings.collect { rendering: RenderingAndSnapshot<String> ->
-          emittedRenderings += rendering.rendering
+  /**
+   * A different form of [onOutput_called_after_rendering_emitted]. Here we launch the workflow
+   * runtime on its own [TestScope] with a [StandardTestDispatcher], which will be paused until
+   * told to advance.
+   *
+   * We *collect* emitted renderings on the [UnconfinedTestDispatcher] of the [runTest].
+   * The point here is that when the runtime sets the value on the StateFlow - a non-suspending
+   * operation - and then it calls [onOutput] - a suspending operation - the [onOutput] handler
+   * will not be immediately dispatched (it is waiting for dispatch from the scheduler), but the
+   * collector of the renderings [StateFlow] will be dispatched and update the 'emitted' renderings.
+   * Then when we let the runtime's scheduler go ahead, it will have already been populated.
+   */
+  @Test fun onOutput_called_after_rendering_emitted_and_collected() {
```

Comment:
Maybe I'm missing something but it looks like what this test is really testing is the behavior of `Unconfined` dispatchers. The behavior it validates is implied by the previous test when the renderings flow is collected with an unconfined dispatcher. Doesn't seem worth writing our own tests for core coroutine features.

Reply:
heh, ya you're point is valid. I was trying to make a point with the test more than actually test anything.